knowledge-asset-tools.js

```diff
@@ -130,7 +136,166 @@ async function peerId2Hash(peerId) {
   return ethers.utils.sha256(ethers.utils.toUtf8Bytes(peerId));
 }
 
+function groupNquadsBySubject(nquadsArray) {
+  const store = new N3.Store();
+  const parser = new N3.Parser({ format: "star" });
+
+  nquadsArray.forEach((quad) => {
+    try {
+      const parsedQuad = parser.parse(quad);
+      parsedQuad.forEach((quad) => store.addQuad(quad));
+    } catch (error) {
+      console.error("Error parsing quad:", quad);
+      console.error(error);
+    }
+  });
+
+  function groupTriples(triples) {
+    const grouped = {};
+
+    triples.forEach((quad) => {
+      let subject = quad.subject ? quad.subject : null;
+      const predicate = quad.predicate.value;
+      let object = quad.object;
+
+      if (subject && subject.termType === "Quad") {
+        const nestedSubjectKey = `<<${subject.subject.value}> <${subject.predicate.value}> ${subject.object.value}>>`;
+
+        if (!grouped[nestedSubjectKey]) {
+          grouped[nestedSubjectKey] = [];
+        }
+
+        grouped[nestedSubjectKey].push([predicate, object.value]);
+      } else {
+        object = object ? object.value : "";
+
+        if (!grouped[subject.value]) {
+          grouped[subject.value] = [];
+        }
+
+        grouped[subject.value].push([predicate, object]);
+      }
+    });
+
+    return grouped;
+  }
+
+  const allTriples = store.getQuads(null, null, null, null);
```

Comment:
we really have to pass 4 times null here? no default arguments? not ur fault if so but im curious if this can be cleaner hahaha